src/nvrtc/safe.rs

```diff
@@ -222,6 +225,257 @@ impl CompileOptions {
     }
 }
 
+#[derive(Debug)]
+enum PtxCrateKind {
+    Cargo{ project_dir: PathBuf },      // rust ptx project dir w/ Cargo.toml & kernel(s) in src/lib.rs
+    _Standalone{ standalone: PathBuf },  // todo!("standalone rs file compiled with rustc")
+}
+
+impl TryFrom<PathBuf> for PtxCrateKind {
+    type Error = String;
+
+    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
+        if !value.exists() {
+            return Err(format!("{value:?} does not exist"))
+        }
+        
+        // if value = path/to/files
+        if let Some(name) = value
+            .file_name()
+            .map(|name| name.to_str())
+        {
+            match name {
+                // value = path/to/project/Cargo.toml
+                Some("Cargo.toml") =>
+                    return Ok( Self::Cargo { project_dir: value.parent().unwrap().into() } ),
+                // value = path/to/project/src/lib.rs
+                Some("lib.rs") => {
+                    let src = value.parent().unwrap();
+                    if let Some(project_dir) = src.parent() {
+                        let manifest = project_dir.join("Cargo.toml");
+                        if manifest.exists() {
+                            return Ok( Self::Cargo { project_dir: project_dir.into() } )
+                        } else {
+                            return Err("standalone RS to PTX not supported".to_string())
+                            // return Ok( Self::Standalone { standalone: value })
+                        }
+                    } else {
+                        return Err(format!("could not find parent of {src:?}"))
+                    }
+                },
+                // value = path/to/project/unsupported_name
+                Some(name) =>
+                    return Err(format!("unsupported file name: {name}")),
+                // name.to_str() failed to parse as unicode
+                None =>
+                    return Err(format!("failed to parse {name:?} as valid unicode"))
+            }
+        }
+
+        // if value = path/to/project/ containing Cargo.toml
+        if value.join("Cargo.toml").exists() {
+            Ok( Self::Cargo { project_dir: value } )
+        } else {
+            Err(format!("{value:?}/Cargo.toml missing"))
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct PtxCrate {
+    project: PtxCrateKind,
+    kernels: Option<Vec<Ptx>>,
+}
+
+impl TryInto<Vec<Ptx>> for PtxCrate {
+    type Error = String;
+    fn try_into(self) -> Result<Vec<Ptx>, Self::Error> {
+        self.kernels.ok_or("kernels not built".into())
+    }
+}
+
+impl TryFrom<PathBuf> for PtxCrate {
+    type Error = String;
+
+    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
+        let project = value.try_into()?;
+        Ok(PtxCrate { project, kernels: None })
+    }
+}
+
+/// `PtxCrate` provides methods to compile a Rust crate to CUDA PTX code,
```

Comment:
Nice doc strings!! These are really helpful

BTW I think you can probably move this to on top of the struct definition instead of the impl block, that's generally the way to do it 😄

Reply:
Hopefully this is better now. I can't `cargo doc` right now because

```bash
  --> src/lib.rs:20:20
   |
20 | //! | cudnn | - | [cudnn::result] | [cudnn::sys] |
   |                    ^^^^^^^^^^^^^ no item named `cudnn` in scope
   |
   = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default

warning: unresolved link to `cudnn::sys`
  --> src/lib.rs:20:38
   |
20 | //! | cudnn | - | [cudnn::result] | [cudnn::sys] |
   |                                      ^^^^^^^^^^ no item named `cudnn` in scope

warning: `cudarc` (lib doc) generated 2 warnings
```